### PR TITLE
8281079: [s390] Unify Address Operand Encoding in Instruction Emitters

### PR DESCRIPTION
**Pull request by RealLucy**
May I please request reviews for this cleanup and unification change. It streamlines the address operand emitters for all instructions. As such, the modifications are widespread and may be cumbersome to review. Please ask questions if anything is unclear.

SAP does not maintain a full s390x build and test environment any longer. I would therefore very much appreciate if somebody would run a build and test cycle on s390x. 

As this is a s390x-only change, other platforms are not impacted.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8281079](https://bugs.openjdk.java.net/browse/JDK-8281079): [s390] Unify Address Operand Encoding in Instruction Emitters

### Reviewers
 * [Martin Doerr](https://openjdk.java.net/census#mdoerr) (@TheRealMDoerr - **Reviewer**) ⚠️ Review applies to 2094825be78559fb0b802a38df62fb31e37ef97a
 * [Matthias Baesken](https://openjdk.java.net/census#mbaesken) (@MBaesken - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7324/head:pull/7324` \
`$ git checkout pull/7324`

Update a local copy of the PR: \
`$ git checkout pull/7324` \
`$ git pull https://git.openjdk.java.net/jdk pull/7324/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7324`

View PR using the GUI difftool: \
`$ git pr show -t 7324`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7324.diff">https://git.openjdk.java.net/jdk/pull/7324.diff</a>

</details>

### CODE REVIEW

**Comment by RealLucy** [comment]
Would not know how s390-only changes could make linux x86 tests fail. I assume it's unrelated.

File: src/hotspot/cpu/s390/assembler_s390.inline.hpp
Comment (TheRealMDoerr) [inline src/hotspot/cpu/s390/assembler_s390.inline.hpp]: Looks correct, but what is the motivation behind splitting unaligned stores? z can store to unaligned memory.

File: src/hotspot/cpu/s390/assembler_s390.inline.hpp
Comment (TheRealMDoerr) [inline src/hotspot/cpu/s390/assembler_s390.inline.hpp]: A bit lengthy for a trivial explanation.

File: src/hotspot/cpu/s390/assembler_s390.inline.hpp
Comment (TheRealMDoerr) [inline src/hotspot/cpu/s390/assembler_s390.inline.hpp]: Tricky, but looks correct.

**Review by TheRealMDoerr — COMMENTED** [review]
Looks basically good. I have a few minor remarks. Did you test this carefully? I can't guarantee that there's no typo or copy&paste bug.

File: src/hotspot/cpu/s390/assembler_s390.hpp
Comment (RealLucy) [reply src/hotspot/cpu/s390/assembler_s390.hpp:1812]: Agreed. Will get removed with next commit.

File: src/hotspot/cpu/s390/assembler_s390.inline.hpp
Comment (RealLucy) [reply src/hotspot/cpu/s390/assembler_s390.inline.hpp]: You mean I should not explain what's going on?

File: src/hotspot/cpu/s390/assembler_s390.inline.hpp
Comment (RealLucy) [reply src/hotspot/cpu/s390/assembler_s390.inline.hpp]: Want some detailed comment here?

File: src/hotspot/cpu/s390/assembler_s390.inline.hpp
Comment (TheRealMDoerr) [reply src/hotspot/cpu/s390/assembler_s390.inline.hpp]: The explanation in the middle is fine. What's the value of the first and last sentence? The first one is already expressed by ShouldNotReachHere, the last one is a general statement which I don't like to be at every switch.

File: src/hotspot/cpu/s390/assembler_s390.inline.hpp
Comment (TheRealMDoerr) [reply src/hotspot/cpu/s390/assembler_s390.inline.hpp]: I can understand it, but a short comment would be appreciated. E.g. length can only be >2 if first 2 bits are != 0.

File: src/hotspot/cpu/s390/assembler_s390.inline.hpp
Comment (RealLucy) [reply src/hotspot/cpu/s390/assembler_s390.inline.hpp]: Will rework with next commit.

File: src/hotspot/cpu/s390/assembler_s390.inline.hpp
Comment (RealLucy) [reply src/hotspot/cpu/s390/assembler_s390.inline.hpp]: Will rework with next commit.

File: src/hotspot/cpu/s390/assembler_s390.inline.hpp
Comment (RealLucy) [reply src/hotspot/cpu/s390/assembler_s390.inline.hpp]: Will go back to single int store for 4-byte instructions.

Comment (TheRealMDoerr) [reply src/hotspot/cpu/s390/assembler_s390.inline.hpp]: Ok. I'm fine with either version. Maybe the reason you did it this way was to be more consistent with the 6 Byte version.

**Review by TheRealMDoerr — APPROVED** [review]
LGTM. Thanks for cleaning it up. Needs to get tested before integration.

**Comment by RealLucy** [comment]
Thank you, Martin, for your review. 
I will perform testing to the best of my abilities.
For formal testing, I have to rely on @backwaterred

**Comment by RealLucy** [comment]
This is a pure s390 change. Test failures on Windows x64 must be unrelated.

**Comment by backwaterred** [comment]
@RealLucy, I am happy to help. I am currently running tier1 tests on s390, and I will update this comment when they complete

---

**Comment by RealLucy** [comment]
> I see a failure in compiler/c2/irTests/TestAutoVectorization2DArray.java. The output is attached. [TestAutoVectorization2DArray.txt](https://github.com/openjdk/jdk/files/8157927/TestAutoVectorization2DArray.txt)

@backwaterred I'm very sure the failure you are seeing is unrelated to my changes. Please note that JEP 338 (Vector API) is not supported on s390. I can reproduce the failure with a plain OpenJDK/jdk build of today.

Could you please verify that you see the failure with a plain build as well? Thank you!

Btw: sorry for reacting with such a long delay. Other duties were keeping me busy.

**Comment by backwaterred** [comment]
> I'm very sure the failure you are seeing is unrelated to my changes. Please note that JEP 338 (Vector API) is not supported on s390. I can reproduce the failure with a plain OpenJDK/jdk build of today.
> 
> Could you please verify that you see the failure with a plain build as well? Thank you!

@RealLucy, I believe you are correct. TestAutoVectorization2DArray.java is failing in the current master branch as well.


> Btw: sorry for reacting with such a long delay. Other duties were keeping me busy.

No problem at all!

**Comment by RealLucy** [comment]
This is a gentle reminder for anybody out there who is willing and able to provide a second review. This is a pure s390 change.

**Comment by RealLucy** [comment]
This PR has been lingering around for 14 days now since my last request for a second review. 
@backwaterred Could you please try to get approval from those in charge that I can integrate this PR? As mentioned above, we have a port of this PR to HotSpot 11 running in our in-house builds and tests since three weeks now - not a single issue. Thanks!

**Comment by RealLucy** [comment]
@backwaterred Sure I can find another SAP reviewer. We have a few in our group. It was my intention to not handle the review process completely SAP-internally. Some people might regard that as too secluded. But if there is no other way...

**Review by MBaesken — APPROVED** [review]

**Comment by RealLucy** [comment]
Thank you for the review, @MBaesken 
I'll go ahead and integrate the PR tomorrow (Apr 02)

**Comment by backwaterred** [comment]
> @backwaterred Sure I can find another SAP reviewer. We have a few in our group. It was my intention to not handle the review process completely SAP-internally. Some people might regard that as too secluded. But if there is no other way...

Ah, that makes sense and is a good idea. It would be ideal to have development be open and cross organizational. Hopefully we can find some fresh 390 reviewers in the future who are not also committed to competing projects.

**Pull request closed without merging**
